src/platform/nrfconnect/wifi/WiFiManager.cpp

```diff
@@ -272,35 +276,91 @@ CHIP_ERROR WiFiManager::GetNetworkStatistics(NetworkStatistics & stats) const
     return CHIP_NO_ERROR;
 }
 
-void WiFiManager::ScanResultClbk(NetEventCallback * cb)
+void WiFiManager::ScanResultHandler(uint8_t * data)
 {
-    VerifyOrReturn(Instance().mScanResultCallback != nullptr);
-    const struct wifi_scan_result * scanResult      = (const struct wifi_scan_result *) cb->info;
-    NetworkCommissioning::WiFiScanResponse response = ToScanResponse(scanResult);
-    Instance().mScanResultCallback(scanResult != nullptr ? &response : nullptr);
+    const struct wifi_scan_result * scanResult = reinterpret_cast<const struct wifi_scan_result *>(data);
+
+    if (Instance().mInternalScan &&
+        strcmp(reinterpret_cast<const char *>(scanResult->ssid), reinterpret_cast<const char *>(Instance().mWantedSSID.data())) ==
+            0)
+    {
+        // Prepare the connection parameters
+        // In case there are many networks with the same SSID choose the one with the best RSSI
+        if (scanResult->rssi > Instance().mWiFiParams.mRssi)
+        {
+            Instance().ClearStationProvisioningData();
+            Instance().mWiFiParams.mParams.ssid_length = Instance().mWantedSSID.size();
+            Instance().mWiFiParams.mParams.ssid        = Instance().mWantedSSID.data();
+            // Fallback to the WIFI_SECURITY_TYPE_PSK if the security is unknown
+            Instance().mWiFiParams.mParams.security =
+                scanResult->security <= WIFI_SECURITY_TYPE_MAX ? scanResult->security : WIFI_SECURITY_TYPE_PSK;
+            Instance().mWiFiParams.mParams.psk_length = Instance().mCurrentCredentials.size();
+
+            // If the security is none, WiFi driver expects the psk to be nullptr
+            if (Instance().mWiFiParams.mParams.security == WIFI_SECURITY_TYPE_NONE)
+            {
+                Instance().mWiFiParams.mParams.psk = nullptr;
+            }
+            else
+            {
+                Instance().mWiFiParams.mParams.psk = const_cast<uint8_t *>(Instance().mCurrentCredentials.data());
+            }
+
+            Instance().mWiFiParams.mParams.timeout = Instance().mHandling.mConnectionTimeoutMs.count();
+            Instance().mWiFiParams.mParams.channel = scanResult->channel;
+            Instance().mWiFiParams.mRssi           = scanResult->rssi;
+        }
+    }
+
+    if (Instance().mScanResultCallback && !Instance().mInternalScan)
+    {
+        NetworkCommissioning::WiFiScanResponse response = ToScanResponse(scanResult);
+        Instance().mScanResultCallback(scanResult != nullptr ? &response : nullptr);
```

Comment:
Is it possible that it's null? Above you seem to assume it is not. In the old API nullptr was used to indicate `done` callbacks, but maybe it's now unnecessary and `mScanResultCallback` could take a reference.

Reply:
Right, I guess reference can be used here. Will fix that.